merlin-framework/src/main/java/gov/nasa/jpl/aerie/merlin/framework/Scoping.java

```diff
@@ -0,0 +1,21 @@
+package gov.nasa.jpl.aerie.merlin.framework;
+
+/**
+ * <p>
+ *   A `Scoping` allows a RootModel to be stashed in a thread-local cell, allowing it to be accessed
+ *   by model methods regardless of their depth in the stack.
+ * </p>
+ * <p>
+ *   The `contextualizeModel` method should always be used in a `try-with-resources` method,
+ *   or in general, in some way enforcing a stack discipline on the thread-local cell. In other words,
+ *   this method emulates *dynamic scoping* in combination with a `try-with-resources` method.
+ * </p>
+ */
+public interface Scoping<Registry, Model> {
+  Undo contextualizeModel(final RootModel<Registry, Model> model);
+
+  // Undos must not throw exceptions, and should be usable in try-with-resources without an Exception catch-all block.
+  interface Undo extends AutoCloseable {
+    void close();
+  }
+}
```

Comment:
Hmm... could you help me understand why we want to use Scoped (and Scoping) for the `model` field of `ActivityTypes`? In general, my understanding of the `Scoped` pattern is that we're setting a global variable to a certain value for the duration of a method call. That way the whole call tree below that sees that value of the global variable. It's structured as a linked list of Undo operations, since it's possible for children to set a new value for their children.

(Writing that out helped it click a lot better for me, wow)

What I still would like clarity on is why we want it for mission models - is it valid for tasks to spawn other tasks based on a different mission model...? (Not even sure I'm asking a well-formed question to begin with)

Reply:
Different threads have different thread-locals (which is how Scoped is implemented). We have to make sure we run each activity with its model in-Scope, or else you'll get ugly exceptions when attempting to spawn child activities.

More generally, I don't want to leave any behind visible side-effects; it's too easy to accidentally rely on the Scoped variable still being set when it shouldn't be. We just want a dynamically-scoped variable, and this is how we maintain the stack discipline for that scoping.

Reply:
The most important reason we're using a `Scoped` cell to store the root model is because, if you're in some method during simulation and you `spawn` some other activity, you need the ID of its activity type that you got when you registered it during model initialization. Those IDs are kept in the RootModel, and we don't want to thread them down through every method call from the activity top-level down to everywhere that needs them -- that's a terrible breach of modularity.